tools/utils.hh

```diff
@@ -61,4 +62,175 @@ const Op& get_selected_operation(int& ac, char**& av, const std::vector<Op>& ope
 // is configured with INFO level.
 void configure_tool_mode(app_template::seastar_options& opts, const sstring& logger_name = {});
 
+class basic_option {
+public:
+    const char* name;
+    const char* description;
+
+public:
+    basic_option(const char* name, const char* description) : name(name), description(description) { }
+
+    virtual void add_option(boost::program_options::options_description& opts) const = 0;
+};
+
+template <typename T = std::monostate>
+class typed_option : public basic_option {
+    std::optional<T> _default_value;
+    int _count;
+
+    virtual void add_option(boost::program_options::options_description& opts) const override {
+        if (_default_value) {
+            opts.add_options()(name, boost::program_options::value<T>()->default_value(*_default_value), description);
+        } else {
+            opts.add_options()(name, boost::program_options::value<T>(), description);
+        }
+    }
+
+public:
+    typed_option(const char* name, const char* description) : basic_option(name, description) { }
+    typed_option(const char* name, T default_value, const char* description) : basic_option(name, description), _default_value(std::move(default_value)) { }
+    typed_option(const char* name, const char* description, int count) : basic_option(name, description), _count(count) { }
+};
+
+template <>
+class typed_option<std::monostate> : public basic_option {
+    virtual void add_option(boost::program_options::options_description& opts) const override {
+        opts.add_options()(name, description);
+    }
+public:
+    typed_option(const char* name, const char* description) : basic_option(name, description) { }
+};
+
+class operation_option {
+    shared_ptr<basic_option> _opt; // need copy to support convenient range declaration of std::vector<option>
+
+public:
+    template <typename T>
+    operation_option(typed_option<T> opt) : _opt(make_shared<typed_option<T>>(std::move(opt))) { }
+
+    const char* name() const { return _opt->name; }
+    const char* description() const { return _opt->description; }
+    void add_option(boost::program_options::options_description& opts) const { _opt->add_option(opts); }
+};
+
+template <typename... Args>
+class operation {
```

Comment:
Let's try to avoid this over-templating.

Reply:
Detemplated in v2.